samples/hybrid-connections/java/simple-websocket-demo/src/main/java/WebsocketListener.java

```diff
@@ -0,0 +1,60 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.Scanner;
+import java.util.concurrent.CompletableFuture;
+import com.microsoft.azure.relay.HybridConnectionListener;
+import com.microsoft.azure.relay.RelayConnectionStringBuilder;
+import com.microsoft.azure.relay.TokenProvider;
+
+public class WebsocketListener {
+	static boolean quit = false;
+
+	static final String CONNECTION_STRING_ENV_VARIABLE_NAME = "RELAY_CONNECTION_STRING";
+	static final RelayConnectionStringBuilder connectionParams = new RelayConnectionStringBuilder(
+			System.getenv(CONNECTION_STRING_ENV_VARIABLE_NAME));
+
+	public static void main(String[] args) throws URISyntaxException {
+		TokenProvider tokenProvider = TokenProvider.createSharedAccessSignatureTokenProvider(
+				connectionParams.getSharedAccessKeyName(), connectionParams.getSharedAccessKey());
+		HybridConnectionListener listener = new HybridConnectionListener(
+				new URI(connectionParams.getEndpoint().toString() + connectionParams.getEntityPath()), tokenProvider);
+
+		listener.openAsync().join();
+		System.out.println("Listener is online. Press ENTER to terminate this program.");
+
+		CompletableFuture.runAsync(() -> {
+			Scanner in = new Scanner(System.in);
+			in.nextLine();
+
+			// Closing the listener will cause it to stop accepting any more connections.
+			// However, it will not shutdown the connections that are already established and still running.
+			listener.close();
+			in.close();
+		});
+
+		while (listener.isOnline()) {
+
+			// If listener closes, then listener.acceptConnectionAsync() will complete with null after closing down
+			listener.acceptConnectionAsync().thenAccept(connection -> {
+				// connection may be null if the listener is closed before receiving a connection
+				if (connection != null) {
+					System.out.println("New session connected.");
+
+					while (connection.isOpen()) {
+						ByteBuffer bytesReceived = connection.readAsync().join();
+						// If the read operation is still pending when connection closes, the read result as null.
+						if (bytesReceived != null) {
+							String msg = new String(bytesReceived.array());
```

Comment:
>String(bytesReceived.array()); [](start = 24, length = 30)

Would this be more accurate?

String(bytesReceived.array(), bytesReceived.arrayOffset(), bytesReceived.remaining());

Reply:
It probably doesn't make the code run any different today, but it could avoid issues in the future if someone in the stack started doing more pooling of arrays.

---
In reply to: [267911787](https://github.com/Azure/azure-relay/pull/41#discussion_r267911787) [](ancestors = 267911787)

Reply:
The whole point of ByteBuffer is so that we can return an byte[]+offset+length. Otherwise we'd just use byte[] alone.
(Frankly, even the byte[] part is technically optional, but I'm not going to worry about having to use byteBuffer.hasArray() to see it it's safe to call .array().)

---
In reply to: [267914318](https://github.com/Azure/azure-relay/pull/41#discussion_r267914318) [](ancestors = 267914318,267911787)